File: 백준/Silver/3273. 두 수의 합/두 수의 합.java
import java.io.*;
import java.util.*;

class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    public static void main(String[] args) throws IOException {
        int n = Integer.parseInt(br.readLine());
        int[] arr = new int[n];
        int result = 0;
        
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        int i = 0;
        while (st.hasMoreTokens()) {
            arr[i] = Integer.parseInt(st.nextToken());
            i++;
        }
        int x = Integer.parseInt(br.readLine());
        
        Arrays.sort(arr);
        
        int left = 0;
        int right = n-1;
        
        while (left < right) {
            int sum = arr[left] + arr[right];
            if (sum == x) {
                result++;
                left++;
                right--;
            }
            else if (sum < x) left++;
            else right--;
        }
        
        bw.write(result + "\n");
        bw.flush();
        bw.close();
        
        
    }
}

File: 백준/Silver/3273. 두 수의 합/두 수의 합.py
import sys
input=sys.stdin.readline

n = int(input())
a = list(map(int, input().split()))
x = int(input())

# n이 10만보다 작거나 같으므로 O(nlogn) 이하로 처리해야함
a.sort()

left = 0
right = len(a) - 1

ans = 0

while left < right:
    left_val = a[left]
    right_val = a[right]
    
    if left_val + right_val == x:
        ans += 1
        left += 1
        right -= 1
    elif left_val + right_val > x:
        right -= 1
    elif left_val + right_val < x:
        left += 1

print(ans)        